### PR TITLE
feat: exclude conversation members from the search results (AR-1706)

### PR DESCRIPTION
**Pull request by trOnk12**
----
#### PR Submission Checklist for internal contributors

- The **PR Title**
  - [ ] conforms to the style of semantic commits messages¹ supported in Wire's Github Workflow²
  - [ ] contains a reference JIRA issue number like `SQPIT-764`
  - [ ] answers the question: _If merged, this PR will: ..._ ³

- The **PR Description**
  - [ ] is free of optional paragraphs and you have filled the relevant parts to the best of your ability
----

# What's new in this PR?

### Issues

- In order to add the participants to the group conversation, we need to search for the users that are not part of the conversation

### Solutions

- create SearchUsersOptions value object, that allows to pass the ConversationID which we want to exclude from the search for users. 
- pass it to searchKnownUsersByNameOrHandleOrEmail, searchKnownUsersByHandle, searchUserDirectory
- create two methods inside userDAO : 
 -  getUsersNotInConversationByHandle
 - getUsersNotInConversationByNameOrHandleOrEmail
- create UserSearchApiWrapper() that wraps the API call to userSearchApi.search and filter the returned results by those that are in the conversation 

### Dependencies (Optional)

_If there are some other pull requests related to this one (e.g. new releases of frameworks), specify them here._

Needs releases with:

- [ ] GitHub link to other pull request

#### Test Coverage (Optional)

- [x] I have added automated test to this contribution

#### How to Test

_Briefly describe how this change was tested and if applicable the exact steps taken to verify that it works as expected._

### Notes (Optional)

_Specify here any other facts that you think are important for this issue._

### Attachments (Optional)

_Attachments like images, videos, etc. (drag and drop in the text box)_

----
#### PR Post Submission Checklist for internal contributors (Optional)

 - [ ] Wire's Github Workflow has automatically linked the PR to a JIRA issue
----
#### PR Post Merge Checklist for internal contributors

 - [ ] If any soft of configuration variable was introduced by this PR, it has been added to the relevant documents and the CI jobs have been updated.
----
##### References
1. https://sparkbox.com/foundry/semantic_commit_messages
1. https://github.com/wireapp/.github#usage
1. E.g. `feat(conversation-list): Sort conversations by most emojis in the title #SQPIT-764`.

### CODE REVIEW

**Comment by trOnk12** [comment]
Make it draft for now as Mohammed introduced me to a SQL query that will do the same as the filtering

File: logic/src/commonTest/kotlin/com/wire/kalium/logic/feature/user/SearchUserDirectoryUseCaseTest.kt [hunk 1]
```diff
@@ -1,103 +1,102 @@
...
+//                        completePicture = null,
+//                        availabilityStatus = UserAvailabilityStatus.NONE,
+//                        userType =  UserType.EXTERNAL
+//                    )
+//                }
+//            }
+//        )
+//    }
+//
+//}
```
Comment (vitorhugods) [inline logic/src/commonTest/kotlin/com/wire/kalium/logic/feature/user/SearchUserDirectoryUseCaseTest.kt:102]: Commented out tests?

File: logic/src/commonTest/kotlin/com/wire/kalium/logic/feature/user/SearchUserUseCaseTest.kt [hunk 2]
```diff
@@ -1,130 +1,130 @@
...
+//                        completePicture = null,
+//                        availabilityStatus = UserAvailabilityStatus.NONE,
+//                        userType = UserType.FEDERATED
+//                    )
+//                }
+//            }
+//        )
+//    }
+//
+//}
```
Comment (vitorhugods) [inline logic/src/commonTest/kotlin/com/wire/kalium/logic/feature/user/SearchUserUseCaseTest.kt:130]: More tests commented out?

**Review by vitorhugods — COMMENTED** [review]
I haven't dived deeper yet, just wanted to point out the commented-out tests.

File: logic/src/commonMain/kotlin/com/wire/kalium/logic/data/publicuser/UserSearchApiWrapper.kt [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+    */
+    suspend fun search(
+        searchQuery: String,
+        domain: String,
+        maxResultSize: Int?,
+        searchUsersOptions: SearchUsersOptions
+    ): Either<NetworkFailure, UserSearchResponse>
+}
+
+class UserSearchApiWrapperImpl(
```
Comment (vitorhugods) [inline logic/src/commonMain/kotlin/com/wire/kalium/logic/data/publicuser/UserSearchApiWrapper.kt:28]:
```suggestion
internal class UserSearchApiWrapperImpl(
```

File: logic/src/commonMain/kotlin/com/wire/kalium/logic/data/publicuser/UserSearchApiWrapper.kt [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+import com.wire.kalium.logic.functional.Either
+import com.wire.kalium.logic.functional.map
+import com.wire.kalium.logic.wrapApiRequest
+import com.wire.kalium.network.api.contact.search.UserSearchApi
+import com.wire.kalium.network.api.contact.search.UserSearchRequest
+import com.wire.kalium.network.api.contact.search.UserSearchResponse
+import com.wire.kalium.persistence.dao.ConversationDAO
+import kotlinx.coroutines.flow.firstOrNull
+
+interface UserSearchApiWrapper {
```
Comment (vitorhugods) [inline logic/src/commonMain/kotlin/com/wire/kalium/logic/data/publicuser/UserSearchApiWrapper.kt:15]: Let's try to make these `internal`, so they don't leak out to consumers of Kalium.

```suggestion
internal interface UserSearchApiWrapper {
```

File: logic/src/commonMain/kotlin/com/wire/kalium/logic/data/publicuser/UserSearchApiWrapper.kt [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+    /*
+    Back-end has not support to return user that are not part of the conversation,
+    therefore we need to filter those users ourself.
+    */
+    suspend fun search(
```
Comment (vitorhugods) [inline logic/src/commonMain/kotlin/com/wire/kalium/logic/data/publicuser/UserSearchApiWrapper.kt:20]: This Kdoc is written more like a comment than actual documentation.

```suggestion
    /*
     * Searches for users that match given the [searchQuery] using the API.
     * Depending on the [searchUsersOptions], the members of a conversation can be excluded.
     */
```

File: logic/src/commonTest/kotlin/com/wire/kalium/logic/feature/user/SearchKnownUserUseCaseTest.kt [hunk 6]
```diff
@@ -1,242 +1,242 @@
...
+//import kotlin.test.Test
+//import kotlin.test.assertIs
+//import com.wire.kalium.logic.data.publicuser.model.UserSearchResult
+//import com.wire.kalium.logic.data.user.UserRepository
+//import com.wire.kalium.logic.feature.publicuser.search.Result
+//import com.wire.kalium.logic.feature.publicuser.search.SearchUsersUseCase
+//import com.wire.kalium.logic.framework.TestUser
+//import kotlin.test.assertFalse
+//
+//class SearchKnownUserUseCaseTest {
```
Comment (vitorhugods) [inline logic/src/commonTest/kotlin/com/wire/kalium/logic/feature/user/SearchKnownUserUseCaseTest.kt:29]: Tests still commented out?

Also, missing tests to assert the new behaviour with `SearchOptions`

File: logic/src/commonTest/kotlin/com/wire/kalium/logic/test_util/TestNetworkException.kt [hunk 7]
```diff
@@ -75,7 +75,8 @@ object TestNetworkException {
 }
 
 object TestNetworkResponseError {
-    fun <T : Any> genericError(): NetworkResponse<T> = NetworkResponse.Error(TestNetworkException.generic)
+    fun genericError(): NetworkFailure = NetworkFailure.NoNetworkConnection(null)
```
Comment (vitorhugods) [inline logic/src/commonTest/kotlin/com/wire/kalium/logic/test_util/TestNetworkException.kt:78]: Not really generic.

```suggestion
    fun noNetworkConnection(): NetworkFailure = NetworkFailure.NoNetworkConnection(null)
```

**Review by vitorhugods — CHANGES_REQUESTED** [review]
Great job. Nice and tidy.

Still missing tests / commenting out tests, and some tiny suggestions.

**Review by vitorhugods — APPROVED** [review]
Nicely done :)

Just a tiny coment for the future, but it's fine.

File: logic/src/commonMain/kotlin/com/wire/kalium/logic/feature/user/UserScope.kt [hunk 8]
```diff
@@ -28,15 +28,15 @@ class UserScope(
     private val syncManager: SyncManager,
     private val assetRepository: AssetRepository,
     private val teamRepository: TeamRepository,
-    private val connectionRepository: ConnectionRepository,
+    private val connectionRepository: ConnectionRepository
```
Comment (vitorhugods) [inline logic/src/commonMain/kotlin/com/wire/kalium/logic/feature/user/UserScope.kt:31]: We can change this `UserScope` to fix the compilation error:

```kt
class UserScope internal constructor(
    private val userRepository: UserRepository,
    private val searchUserRepository: SearchUserRepository,
    private val syncManager: SyncManager,
    private val assetRepository: AssetRepository,
    private val teamRepository: TeamRepository,
    private val connectionRepository: ConnectionRepository
)
```

Comment (trOnk12) [reply logic/src/commonMain/kotlin/com/wire/kalium/logic/feature/user/UserScope.kt:31]: done

File: logic/src/commonTest/kotlin/com/wire/kalium/logic/feature/user/SearchKnownUserUseCaseTest.kt [hunk 9]
```diff
@@ -137,9 +140,81 @@ class SearchKnownUserUseCaseTest {
...
+        verify(arrangement.searchUserRepository)
+            .suspendFunction(arrangement.searchUserRepository::searchKnownUsersByNameOrHandleOrEmail)
+            .with(anything(), eq(searchUsersOptions))
+            .wasInvoked(exactly = once)
+    }
+
 }
 
-class Arrangement {
+internal class Arrangement {
```
[the remark this answers is not in the document]
Comment (trOnk12) [reply logic/src/commonTest/kotlin/com/wire/kalium/logic/feature/user/SearchKnownUserUseCaseTest.kt:217]: done

**Pull request merged**